### DISCUSSION TITLE
Consistent string value for BodyID?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#113] [Q&A] [answered]
One of the oddly difficult parts my brain is struggling with is things like  `BodyID` are WebAssembly objects and not a simple string etc. 

When trying to build an API for systems or more specifically to manage JavaScript Worker communications I want to send commands or get information about a body and it would be easiest to do this with a simple string, and I think impossible with the the current `BodyID` object.

[In Rapier, this is done with the ](https://rapier.rs/docs/user_guides/javascript/rigid_bodies/#creation-and-insertion)`RidgedBody.handle`

In going over [the Jolt Docs, I saw](https://jrouwe.github.io/JoltPhysics/class_body_i_d.html) it's possible to create a `BodyID` with a specified ID value, and I saw in the examples using  `<BodyID>.GetIndexAndSequenceNumber();`

However in reading through the docs it seems like the Body Index and Sequence values may change when bodies are added/removed.

I think I understand the sequence, where two bodies may have the same index value but as they were created at different points the sequence is used to to determine if they are the same. 

But I'm not sure if the bodies index changes as other bodies are added/removed. 

In Javascript, if a previous sibling in an Array is removed, the current item's index lowers. I'm not sure if Jolt's Bodies have similar behaviors...

Does the `GetIndexAndSequenceNumber()` value stay consistent throughout the bodies life?

**Comment by LeXXik** [comment]
In my setup, I maintain a map that maps my own IDs with the bodies in the world. Something like this or a variation of it, depending on your use case:

```js
const idxMap = new Map();
const bodyMap = new Map();

idxMap.set(Jolt.getPointer(body), customIndex);
bodyMap.set(customIndex, body);
```

In this case `customIndex` is an integer that I send from main thread to web worker. Main thread is not aware of actual Jolt's body IDs or any of its other implementations.

**Reply by DennisSmolek** [reply]
I'm doing something similar. I wrap the bodies in another object before mapping and generate my own keys. 

However if Jolt is already generating keys for the bodies I can make sure they all match correctly. Especially when the kinks get worked out with multi-threading.

Sidenote, I'm mostly storing the `BodyID` and getting values with the `BodyInterface` but have considered storing the bodies directly like that. It the `Jolt.getPointer(body)` required for body objects as well?

**Reply by LeXXik** [reply]
`Jolt.getPointer(body)` will equal to zero if the body is no longer valid. It is not strictly necessary, but I found it useful in some scenarios of my setup when running it in worker over querying an ID. Some callbacks, e.g. during a raycast, can receive a body, which was already removed from simulation.

**Comment by jrouwe** [comment, marked as answer]
> Does the GetIndexAndSequenceNumber() value stay consistent throughout the bodies life?

Yes it does. And you can convert it back and forth to a BodyID:

```
let bodyId = ...
let number = bodyId.GetIndexAndSequenceNumber();
let recoveredBodyId = new Jolt.BodyID(number);
```

(note that I had to make a small change to the API to make this possible, I had forgotten to expose this constructor)
